File: internal/bus/bus.go
// Package bus turns the events table into live delivery.
//
// The invariant everything here is built around, and the one a future
// contributor will violate:
//
//	The events table is the transport. NOTIFY is a wakeup bell carrying an id.
//	Every consumer stays correct if every notification is dropped.
//
// So the tailer never trusts a notification. It re-reads an overlap window on
// every cycle, dedupes by id, and polls unconditionally on a timer whether or
// not the listening connection is healthy. A missed notification is a latency
// event, never a correctness event ... and the tests prove that by running the
// whole suite with notifications disabled.
package bus

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"math/rand/v2"
	"sync"
	"sync/atomic"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/jackc/pgxlisten"

	"github.com/bees-roadhouse/hive-sandbox/internal/store"
)

// Config tunes the tailer. The defaults are the ones the design argues for;
// every one of them is a latency or safety knob, not a performance knob.
type Config struct {
	// Overlap is how far back every poll re-reads.
	//
	// This is the load-bearing number. bigserial ids are assigned BEFORE
	// commit, so a row assigned early and committed late becomes visible after
	// rows with higher ids. A tail that advanced past it would skip it forever.
	// Overlap must exceed the longest transaction that writes events.
	Overlap time.Duration

	// PollInterval is the unconditional backstop, run regardless of connection
	// health. It is what turns a missed notification into a latency event.
	PollInterval time.Duration

	// BatchLimit bounds one tail query. A full batch means "there is more",
	// and the loop goes straight round again rather than waiting for the timer.
	BatchLimit int

	// ReconnectDelay for the listening connection. pgxlisten defaults to ONE
	// MINUTE, which would make a brief database blip look like an hour of
	// silence; 1-2 seconds with jitter is the design's number.
	ReconnectDelay time.Duration

	// Channel is the one coarse channel to listen on. Overriding it is how the
	// test suite proves the invariant that gives this package its shape: point
	// a bus at a channel nobody notifies, and every consumer must still be
	// correct on the backstop poll alone.
	Channel string

	Logger *slog.Logger
}

func (c Config) withDefaults() Config {
	if c.Overlap <= 0 {
		c.Overlap = 5 * time.Second
	}
	if c.PollInterval <= 0 {
		c.PollInterval = 5 * time.Second
	}
	if c.BatchLimit <= 0 {
		c.BatchLimit = 500
	}
	if c.ReconnectDelay <= 0 {
		c.ReconnectDelay = 1500 * time.Millisecond
	}
	if c.Channel == "" {
		c.Channel = store.NotifyChannel
	}
	if c.Logger == nil {
		c.Logger = slog.Default()
	}
	return c
}

// Bus owns one listening connection per host and fans out in memory to every
// subscriber on that host (D4.8). Never one listening connection per client:
// the cost model is a signal per listening backend, so listener count is the
// number to keep small.
type Bus struct {
	pool *pgxpool.Pool
	cfg  Config
	hub  *hub

	// wake carries "something may have happened". It is capacity 1 and
	// non-blocking to send: a notification is a hint, and dropping a redundant
	// one costs nothing.
	wake chan struct{}

	// settled is the newest cursor position that can no longer gain rows
	// behind it, in unix micros. Subscribers checkpoint here rather than at the
	// newest event they saw; see sse.go.
	settledMicros atomic.Int64

	// notified counts notifications actually received, so a test can assert it
	// stayed correct with zero.
	notified atomic.Int64
	polls    atomic.Int64

	startOnce sync.Once
	ready     chan struct{}
}

// New builds a bus over an existing pool. Nothing runs until Run is called.
func New(pool *pgxpool.Pool, cfg Config) *Bus {
	return &Bus{
		pool:  pool,
		cfg:   cfg.withDefaults(),
		hub:   newHub(),
		wake:  make(chan struct{}, 1),
		ready: make(chan struct{}),
	}
}

// Settled is the watermark a subscriber may safely resume from: every event at
// or before it has been read, and no transaction can still commit one behind
// it. Zero until the first tail cycle completes.
func (b *Bus) Settled() time.Time {
	micros := b.settledMicros.Load()
	if micros == 0 {
		return time.Time{}
	}
	// A zero time round-trips through UnixMicro as a zero time rather than as
	// 0, so this covers the empty-log case as well as the never-set one.
	return time.UnixMicro(micros).UTC()
}

// Stats reports how the tailer has been woken. A healthy system polls
// occasionally and is notified often; a system with a dead listener polls only,
// stays correct, and gets slower. Both are visible here.
func (b *Bus) Stats() (notifications, polls int64) {
	return b.notified.Load(), b.polls.Load()
}

// Ready closes once the first tail cycle has run, so callers do not race the
// initial watermark.
func (b *Bus) Ready() <-chan struct{} { return b.ready }

// Run drives the listener and the tail loop until ctx is cancelled.
func (b *Bus) Run(ctx context.Context) error {
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		b.listen(ctx)
	}()

	err := b.tailLoop(ctx)
	wg.Wait()
	b.hub.closeAll()
	return err
}

// listen keeps a DEDICATED connection subscribed to the one coarse channel.
//
// Never from the pool: the pool can hand the connection to someone else or
// reset it, and the subscription dies silently. pgxlisten takes ownership of
// whatever Connect returns, which is why this acquires a standalone conn rather
// than borrowing one.
func (b *Bus) listen(ctx context.Context) {
	listener := &pgxlisten.Listener{
		Connect: func(ctx context.Context) (*pgx.Conn, error) {
			return pgx.ConnectConfig(ctx, b.pool.Config().ConnConfig.Copy())
		},
		LogError: func(ctx context.Context, err error) {
			if ctx.Err() != nil {
				return
			}
			// Non-fatal by design. The backstop poll is already covering us,
			// which is the entire reason a listener outage is survivable.
			b.cfg.Logger.Warn("bus listener", "err", err)
		},
		// Jitter so N hosts do not reconnect in lockstep after a restart.
		// Jitter so N hosts do not reconnect in lockstep. Nothing here is a
		// secret, so a weak generator is the right one.
		ReconnectDelay: b.cfg.ReconnectDelay + rand.N(500*time.Millisecond), //nolint:gosec // spreading reconnects, not generating a key
	}

	listener.Handle(b.cfg.Channel, pgxlisten.HandlerFunc(
		func(_ context.Context, _ *pgconn.Notification, _ *pgx.Conn) error {
			// Never block here, and never query on the listening connection.
			// Buffer and return; the tail loop does the reading.
			b.notified.Add(1)
			b.kick()
			return nil
		}))

	if err := listener.Listen(ctx); err != nil && ctx.Err() == nil {
		b.cfg.Logger.Error("bus listener stopped", "err", err)
	}
}

func (b *Bus) kick() {
	select {
	case b.wake <- struct{}{}:
	default:
	}
}

// tailLoop is the only reader of the events table for live delivery.
func (b *Bus) tailLoop(ctx context.Context) error {
	// Start at the head: a subscriber replays its own history from its own
	// cursor, so there is nothing to gain from pushing the whole log through
	// the hub at boot.
	head, err := store.Head(ctx, b.pool)
	if err != nil {
		return fmt.Errorf("bus: read head: %w", err)
	}

	t := &tailer{
		bus:    b,
		cursor: head,
		seen:   map[int64]time.Time{},
	}
	if head.ID != 0 {
		t.seen[head.ID] = head.At
	}

	ticker := time.NewTicker(b.cfg.PollInterval)
	defer ticker.Stop()

	for {
		if err := t.cycle(ctx); err != nil {
			if ctx.Err() != nil {
				return nil
			}
			// A failed cycle is not fatal. The next one re-reads the same
			// window, so nothing is lost by carrying on.
			b.cfg.Logger.Warn("bus tail cycle", "err", err)
		}
		b.startOnce.Do(func() { close(b.ready) })

		select {
		case <-ctx.Done():
			return nil
		case <-ticker.C:
			b.polls.Add(1)
		case <-b.wake:
		}
	}
}

// tailer holds the cursor state that makes the overlap rule work.
type tailer struct {
	bus *Bus

	// cursor is the highest (created_at, id) this tailer has READ, which is not
	// the same as the highest it has delivered.
	//
	// A burst larger than one batch used to wedge live delivery permanently: the
	// query rewound to `newest delivered minus the overlap` and returned the
	// OLDEST rows in that window, so once those were all deduped nothing was
	// fresh, the position never moved, the batch was still full, and the loop
	// ran the identical query against Postgres forever delivering nothing.
	//
	// What fixes that is store.Tail being a strict forward cursor rather than a
	// rewind, which is measured: reverting only the query wedges the burst test,
	// reverting only this advance-on-read does not. Advancing on rows READ is
	// kept because it is what makes the cursor mean "the log has been read to
	// here" rather than "something interesting was found here", and a position
	// that skips over duplicates is a position that lies about progress.
	//
	// Catching late commits is a separate bounded sweep, below.
	cursor store.Cursor

	// seen is what makes re-reading the window free of duplicates. Keyed by id
	// because that is the only thing guaranteed unique, and pruned by EVENT time
	// relative to the cursor rather than by wall clock.
	//
	// The two used to disagree, and the disagreement had a second face: caught
	// up they agreed and a full duplicate batch livelocked, but far behind, wall
	// clock pruning had already dropped ids the next query would re-read, so
	// nothing looked duplicate and the loop escaped by re-delivering the overlap
	// window on every batch. One defect, livelock or duplicate delivery
	// depending only on how old the backlog was, and neither errored.
	seen map[int64]time.Time
}

// sweepFactor bounds the late-commit sweep relative to a normal batch. The
// window is only as wide as the longest event-writing transaction, so it should
// hold far fewer rows than a batch; the multiplier is headroom, and filling it
// is reported rather than silently truncated.
const sweepFactor = 4

func (t *tailer) cycle(ctx context.Context) error {
	cfg := t.bus.cfg

	// One clock read per cycle rather than per iteration. Under a backlog the
	// inner loop runs many times, and this was half the round trips.
	dbNow, err := store.Now(ctx, t.bus.pool)
	if err != nil {
		return err
	}

	// 1. Drain forward. Strictly after the cursor, so every full batch moves.
	for {
		events, err := store.Tail(ctx, t.bus.pool, t.cursor, cfg.BatchLimit)
		if err != nil {
			return err
		}
		t.deliver(dbNow, events)

		if len(events) < cfg.BatchLimit {
			break
		}
		if err := ctx.Err(); err != nil {
			return err
		}
	}

	// 2. Sweep the window behind the cursor for anything that committed late.
	//
	// bigserial ids are assigned before commit, so a transaction that took its
	// id early and committed just now sits BELOW the cursor and step 1 will
	// never return it. This is the only thing that catches it, and dropping it
	// would reintroduce the bug the overlap exists for.
	if !t.cursor.Zero() {
		from := t.cursor.At.Add(-cfg.Overlap)
		limit := cfg.BatchLimit * sweepFactor
		late, err := store.TailWindow(ctx, t.bus.pool, from, t.cursor, limit)
		if err != nil {
			return err
		}
		if len(late) == limit {
			// Truncating this sweep silently is how a late commit gets lost,
			// which is the exact failure mode the sweep exists to prevent.
			cfg.Logger.Warn("bus: late-commit sweep filled its limit; overlap window may be too wide",
				"limit", limit, "overlap", cfg.Overlap)
		}
		t.deliver(dbNow, late)
	}

	t.updateSettled(dbNow)
	t.prune()
	return nil
}

// updateSettled publishes the watermark a subscriber may safely resume from.
//
// Clamped to the cursor, because a watermark ahead of what has actually been
// read would invite a client to checkpoint past events nobody has delivered.
// While the log is empty the cursor is the zero time and so is the watermark,
// which Settled reports as "not yet known" rather than as the epoch.
func (t *tailer) updateSettled(dbNow time.Time) {
	settled := dbNow.Add(-t.bus.cfg.Overlap)
	if settled.After(t.cursor.At) {
		settled = t.cursor.At
	}
	t.bus.settledMicros.Store(settled.UnixMicro())
}

// deliver advances the cursor over everything read, publishes the watermark,
// then broadcasts whatever was not already delivered.
//
// The watermark moves BEFORE the broadcast on purpose. A subscriber that has
// just been handed an event must not be able to read a watermark that predates
// its own arrival ... otherwise it checkpoints behind where it actually is,
// which is only ever seen as a duplicate after a reconnect and is therefore
// exactly the kind of thing nobody notices.
func (t *tailer) deliver(dbNow time.Time, events []store.Event) {
	fresh := make([]store.Event, 0, len(events))
	for _, e := range events {
		// The cursor moves for every row READ. A duplicate still proves the
		// log reached this position.
		if t.cursor.Before(e.Cursor()) {
			t.cursor = e.Cursor()
		}
		if _, dup := t.seen[e.ID]; dup {
			continue
		}
		t.seen[e.ID] = e.CreatedAt
		fresh = append(fresh, e)
	}
	t.updateSettled(dbNow)
	if len(fresh) > 0 {
		t.bus.hub.broadcast(fresh)
	}
}

// dedupeFloor is the shortest an id stays in the dedupe set.
//
// The natural retention is twice the overlap, since an entry has to outlive
// every window that can still re-read its row. The floor exists because a very
// small overlap would collapse the dedupe set to nothing and re-deliver rows
// forever ... found by a test that configured a near-zero overlap and watched
// the same event arrive twice.
const dedupeFloor = time.Minute

// prune keeps the dedupe set bounded.
func (t *tailer) prune() {
	pruneSeen(t.seen, t.cursor.At, 2*t.bus.cfg.Overlap)
}

// pruneSeen drops ids that can no longer come back, and it is one function
// because the mistake it prevents was made twice in this package at two layers.
//
// The map's values are DATABASE time. `now` must be database time too ... the
// position the reader has reached in the log, not the host's wall clock. The
// tailer passes its cursor; the SSE stream passes the newest event it has
// delivered. Comparing database timestamps against time.Now() is comparing two
// clocks that are allowed to disagree, and it fails in opposite directions
// depending only on how far behind the reader is: caught up they agree and
// nothing looks wrong, far behind the entries are evicted early and every row
// the overlap re-reads arrives a second time.
//
// A zero `now` ... a reader that has established no position yet ... needs no
// special case, and this says so rather than carrying a guard for it. The
// cutoff would land a retain window before year 1, which no database timestamp
// precedes, so nothing is dropped. That branch was written first; a mutation
// run showed the test covering it could not fail, because there is no input
// that distinguishes having it from not.
func pruneSeen(seen map[int64]time.Time, now time.Time, retain time.Duration) {
	if retain < dedupeFloor {
		retain = dedupeFloor
	}
	cutoff := now.Add(-retain)
	for id, at := range seen {
		if at.Before(cutoff) {
			delete(seen, id)
		}
	}
}

// Subscribe joins the in-memory fan-out. Close the subscription when done.
func (b *Bus) Subscribe(buffer int) *Subscription { return b.hub.subscribe(buffer) }

// ErrSlowConsumer means a subscriber fell far enough behind that the hub
// dropped it rather than let it hold memory for everyone. The subscriber should
// reconnect and replay from its last cursor, which is exactly what an SSE client
// does anyway.
var ErrSlowConsumer = errors.New("subscriber fell behind")

File: crates/hive-bus/tests/bus.rs
//! The invariant this crate exists to hold, and the fan-out. Ported from
//! bus_test.go; each test names its origin.

use std::collections::HashSet;
use std::time::Duration;

use hive_bus::{Bus, Config};
use hive_identity::{Credential, Owner, PrincipalKind};
use hive_store::{BootstrapConfig, Event, Store, append_events};
use hive_testdb::TestDb;
use sqlx::PgPool;
use tokio_util::sync::CancellationToken;
use uuid::Uuid;

struct Harness {
    db: TestDb,
    store: Store,
    alice: Uuid,
    cred: Credential,
    cancel: CancellationToken,
    runs: Vec<tokio::task::JoinHandle<()>>,
}

impl Harness {
    async fn new(test: &str) -> Option<Harness> {
        let db = TestDb::new(test).await?;
        hive_store::migrate(db.pool()).await.expect("migrate");
        let mut conn = db.pool().acquire().await.unwrap();
        hive_store::ensure_event_partitions(&mut conn, 1)
            .await
            .expect("partitions");
        drop(conn);
        let store = Store::from_pool(db.pool().clone());
        let res = store
            .bootstrap_in_tx(&BootstrapConfig {
                root_handle: "alice".into(),
                root_name: "Alice".into(),
                ..Default::default()
            })
            .await
            .expect("bootstrap");
        let alice = res.root_actor_id;
        Some(Harness {
            db,
            store,
            alice,
            cred: Credential::new(alice, PrincipalKind::User, alice),
            cancel: CancellationToken::new(),
            runs: Vec::new(),
        })
    }

    fn pool(&self) -> &PgPool {
        self.db.pool()
    }

    fn owner(&self) -> Owner {
        Owner::user(self.alice)
    }

    /// Starts a bus and waits until it has tailed once.
    async fn run(&mut self, cfg: Config) -> Bus {
        let b = Bus::new(self.pool().clone(), cfg);
        let bus = b.clone();
        let cancel = self.cancel.clone();
        self.runs
            .push(tokio::spawn(async move { bus.run(cancel).await }));
        tokio::time::timeout(Duration::from_secs(10), b.ready())
            .await
            .expect("bus never became ready");
        b
    }

    async fn stop(mut self) {
        self.cancel.cancel();
        for r in self.runs.drain(..) {
            let _ = tokio::time::timeout(Duration::from_secs(10), r).await;
        }
    }

    async fn human(&self, handle: &str) -> Uuid {
        let id = Uuid::new_v4();
        sqlx::query(
            "INSERT INTO actors (id, kind, handle, display_name, principal_kind, principal_id, created_by_actor)
             VALUES ($1, 'human', $2, $2, 'user', $1, $3)",
        )
        .bind(id)
        .bind(handle)
        .bind(self.alice)
        .execute(self.pool())
        .await
        .unwrap();
        id
    }

    async fn append(&self, kind: &str, owner: Owner) -> Event {
        let mut ev = Event::new(
            kind,
            &self.cred,
            format!("{{\"kind\":{kind:?}}}").into_bytes(),
        );
        ev.owner = owner;
        let mut conn = self.pool().acquire().await.unwrap();
        append_events(&mut conn, std::slice::from_mut(&mut ev))
            .await
            .expect("append");
        ev
    }

    /// Rings the wakeup bell on a channel of this test's own.
    async fn notify(&self, channel: &str) {
        sqlx::query("SELECT pg_notify($1, '')")
            .bind(channel)
            .execute(self.pool())
            .await
            .unwrap();
    }
}

/// Drains a subscription until it has n events or the deadline passes.
async fn collect(sub: &mut hive_bus::Subscription, n: usize, within: Duration) -> Vec<Event> {
    let deadline = tokio::time::Instant::now() + within;
    let mut got = Vec::new();
    while got.len() < n {
        match tokio::time::timeout_at(deadline, sub.recv()).await {
            Ok(Some(batch)) => got.extend(batch),
            Ok(None) | Err(_) => break,
        }
    }
    got
}

fn quiet(poll: Duration, overlap: Duration) -> Config {
    Config {
        channel: "a_channel_nobody_notifies".into(),
        poll_interval: poll,
        overlap,
        ..Config::default()
    }
}

/// Ported from `TestCorrectWithEveryNotificationDropped`: THE test. If this
/// fails, "a missed notification is a latency event, never a correctness
/// event" is false.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn correct_with_every_notification_dropped() {
    let Some(mut h) = Harness::new("bus_correct_without_notifications").await else {
        return;
    };
    // A batch limit under the event count, so the inner catch-up loop runs.
    let b = h
        .run(Config {
            batch_limit: 8,
            ..quiet(Duration::from_millis(150), Duration::from_secs(2))
        })
        .await;
    let mut sub = b.subscribe(256);
    let mut want = HashSet::new();
    for i in 0..30 {
        want.insert(h.append(&format!("test.event.{i}"), h.owner()).await.id);
    }
    let got = collect(&mut sub, 30, Duration::from_secs(15)).await;
    assert_eq!(
        got.len(),
        30,
        "received {} of 30 events with notifications dropped",
        got.len()
    );
    for e in &got {
        want.remove(&e.id);
    }
    assert!(want.is_empty(), "{} events never arrived", want.len());
    let (notifications, polls) = b.stats();
    assert_eq!(
        notifications, 0,
        "the bus received notifications; the test did not prove what it claims"
    );
    assert!(polls > 0, "the backstop poll never ran");
    h.stop().await;
}

/// Ported from `TestBurstLargerThanOneBatchIsFullyDelivered`.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn burst_larger_than_one_batch_is_fully_delivered() {
    let Some(mut h) = Harness::new("bus_burst").await else {
        return;
    };
    let b = h
        .run(Config {
            batch_limit: 10,
            ..quiet(Duration::from_millis(100), Duration::from_secs(5))
        })
        .await;
    let mut sub = b.subscribe(1024);
    // One call, so every row lands inside one overlap window.
    let mut events: Vec<Event> = (0..120)
        .map(|i| Event::new(format!("burst.{i}"), &h.cred, b"{}".to_vec()))
        .collect();
    let mut conn = h.pool().acquire().await.unwrap();
    append_events(&mut conn, &mut events)
        .await
        .expect("append burst");
    drop(conn);
    let mut want: HashSet<i64> = events.iter().map(|e| e.id).collect();
    for e in collect(&mut sub, 120, Duration::from_secs(30)).await {
        want.remove(&e.id);
    }
    assert!(
        want.is_empty(),
        "{} of 120 events never arrived; the tailer stopped making progress",
        want.len()
    );
    // And it is still alive afterwards: the wedge's signature was silence.
    let after = h.append("after.the.burst", h.owner()).await;
    let tail = collect(&mut sub, 1, Duration::from_secs(20)).await;
    assert!(
        tail.len() == 1 && tail[0].id == after.id,
        "nothing was delivered after the burst; the tailer is wedged"
    );
    let (_, polls) = b.stats();
    assert!(polls > 0, "the tailer never returned to its poll loop");
    h.stop().await;
}

/// A transaction holding an unfinished event insert.
struct LateEvent {
    conn: sqlx::pool::PoolConnection<sqlx::Postgres>,
    id: i64,
}

async fn begin_late_event(h: &Harness) -> LateEvent {
    let mut conn = h.pool().acquire().await.unwrap();
    sqlx::query("BEGIN").execute(&mut *conn).await.unwrap();
    let id: i64 = sqlx::query_scalar(
        "INSERT INTO events (kind, owner_kind, owner_id, author_actor, principal_kind, principal_id)
         VALUES ('late.event', 'user', $1, $1, 'user', $1) RETURNING id",
    )
    .bind(h.alice)
    .fetch_one(&mut *conn)
    .await
    .unwrap();
    LateEvent { conn, id }
}

impl LateEvent {
    async fn commit(mut self) -> i64 {
        sqlx::query("COMMIT")
            .execute(&mut *self.conn)
            .await
            .unwrap();
        self.id
    }
}

/// Ported from `TestLateCommitWithALowerIDIsNotSkipped`: the bug the overlap
/// window exists for.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn late_commit_with_a_lower_id_is_not_skipped() {
    let Some(mut h) = Harness::new("bus_late_commit").await else {
        return;
    };
    let b = h
        .run(quiet(Duration::from_millis(100), Duration::from_secs(5)))
        .await;
    let mut sub = b.subscribe(64);
    let late = begin_late_event(&h).await;
    let fast = h.append("fast.event", h.owner()).await;
    let got = collect(&mut sub, 1, Duration::from_secs(10)).await;
    assert!(
        got.len() == 1 && got[0].id == fast.id,
        "expected the fast event first, got {got:?}"
    );
    let late_id = late.commit().await;
    assert!(late_id < fast.id, "the test is not reproducing the hazard");
    let got = collect(&mut sub, 1, Duration::from_secs(10)).await;
    assert!(
        got.len() == 1 && got[0].id == late_id,
        "the late-committing row (id {late_id}) was never delivered; got {got:?}"
    );
    h.stop().await;
}

/// Ported from `TestLateCommitIsSkippedWithoutTheOverlap`: the negative
/// control.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn late_commit_is_skipped_without_the_overlap() {
    let Some(mut h) = Harness::new("bus_late_commit_no_overlap").await else {
        return;
    };
    let b = h
        .run(quiet(Duration::from_millis(100), Duration::from_nanos(1)))
        .await;
    let mut sub = b.subscribe(64);
    let late = begin_late_event(&h).await;
    let fast = h.append("fast.event", h.owner()).await;
    let got = collect(&mut sub, 1, Duration::from_secs(10)).await;
    assert!(got.len() == 1 && got[0].id == fast.id);
    let late_id = late.commit().await;
    // A near-zero overlap also collapses the dedupe window, so already
    // delivered rows may come round again; only the LATE id matters.
    for e in collect(&mut sub, 50, Duration::from_secs(2)).await {
        assert_ne!(
            e.id, late_id,
            "with the overlap disabled the late row still arrived; the overlap is not what makes the positive test pass"
        );
    }
    h.stop().await;
}

/// Ported from `TestNotificationDeliversFasterThanThePoll`.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn notification_delivers_faster_than_the_poll() {
    let Some(mut h) = Harness::new("bus_notification_fast").await else {
        return;
    };
    // Its own channel: a NOTIFY channel is database-wide, and any other test
    // appending an event would satisfy the warmup for an unrelated reason.
    let channel = format!("bus_test_{}", Uuid::new_v4().simple());
    let b = h
        .run(Config {
            channel: channel.clone(),
            poll_interval: Duration::from_secs(30),
            overlap: Duration::from_secs(2),
            ..Config::default()
        })
        .await;
    let mut sub = b.subscribe(256);
    // The listener connects asynchronously. Warm up until a notification has
    // actually been received.
    let warmup = tokio::time::Instant::now() + Duration::from_secs(20);
    while b.stats().0 == 0 {
        assert!(
            tokio::time::Instant::now() < warmup,
            "the listener never subscribed"
        );
        h.append("warmup", h.owner()).await;
        h.notify(&channel).await;
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    let (before, _) = b.stats();
    let start = tokio::time::Instant::now();
    let want = h.append("journal.entry.created", h.owner()).await;
    h.notify(&channel).await;
    let deadline = start + Duration::from_secs(5);
    let elapsed = loop {
        match tokio::time::timeout_at(deadline, sub.recv()).await {
            Ok(Some(batch)) => {
                if batch.iter().any(|e| e.id == want.id) {
                    break start.elapsed();
                }
            }
            Ok(None) => panic!("subscription closed"),
            Err(_) => panic!("the event never arrived within 5s of a 30s poll interval"),
        }
    };
    assert!(
        elapsed <= Duration::from_secs(2),
        "delivery took {elapsed:?} with a 30s poll interval"
    );
    assert!(
        b.stats().0 > before,
        "delivered without a new notification, so this test proves nothing"
    );
    h.stop().await;
}

/// Ported from `TestVisibilityIsPerSubscriber` (D4.9).
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn visibility_is_per_subscriber() {
    let Some(mut h) = Harness::new("bus_visibility_per_subscriber").await else {
        return;
    };
    let bob = h.human("bob").await;
    let bob_cred = Credential::new(bob, PrincipalKind::User, bob);
    let b = h
        .run(Config {
            poll_interval: Duration::from_millis(200),
            overlap: Duration::from_secs(2),
            ..Config::default()
        })
        .await;
    let mut sub = b.subscribe(64);
    let mine = h.append("alice.private", h.owner()).await;
    let theirs = h.append("bob.private", Owner::user(bob)).await;
    let got = collect(&mut sub, 2, Duration::from_secs(10)).await;
    assert_eq!(got.len(), 2, "the hub is unfiltered by design");
    let g = h.store.guard();
    let mut conn = h.pool().acquire().await.unwrap();
    let for_alice = g.visible(&mut conn, &h.cred, &got).await.unwrap();
    assert!(
        for_alice.len() == 1 && for_alice[0].id == mine.id,
        "alice saw {} events",
        for_alice.len()
    );
    let for_bob = g.visible(&mut conn, &bob_cred, &got).await.unwrap();
    assert!(
        for_bob.len() == 1 && for_bob[0].id == theirs.id,
        "bob saw {} events",
        for_bob.len()
    );
    drop(conn);
    h.stop().await;
}

/// Ported from `TestSlowSubscriberIsDroppedNotBlocking`.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn slow_subscriber_is_dropped_not_blocking() {
    let Some(mut h) = Harness::new("bus_slow_subscriber").await else {
        return;
    };
    let b = h
        .run(Config {
            poll_interval: Duration::from_millis(100),
            overlap: Duration::from_secs(2),
            ..Config::default()
        })
        .await;
    let mut slow = b.subscribe(1);
    let mut healthy = b.subscribe(64);
    for i in 0..8 {
        h.append(&format!("flood.{i}"), h.owner()).await;
        tokio::time::sleep(Duration::from_millis(120)).await;
    }
    // Never read from `slow` until now: the buffered batch, then None.
    let dropped = tokio::time::timeout(Duration::from_secs(10), async {
        loop {
            if slow.recv().await.is_none() {
                break;
            }
        }
    })
    .await;
    assert!(
        dropped.is_ok(),
        "a subscriber that never read was not dropped"
    );
    let got = collect(&mut healthy, 8, Duration::from_secs(10)).await;
    assert_eq!(
        got.len(),
        8,
        "the healthy subscriber fell behind while another was stuck"
    );
    h.stop().await;
}

/// Ported from `TestSettledWatermarkLagsTheNewestEvent`.
#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
async fn settled_watermark_lags_the_newest_event() {
    let Some(mut h) = Harness::new("bus_settled_lags").await else {
        return;
    };
    let b = h
        .run(Config {
            poll_interval: Duration::from_millis(100),
            overlap: Duration::from_secs(3),
            ..Config::default()
        })
        .await;
    let mut sub = b.subscribe(16);
    let e = h.append("journal.entry.created", h.owner()).await;
    assert_eq!(collect(&mut sub, 1, Duration::from_secs(10)).await.len(), 1);
    let settled = b.settled().expect("no watermark after a delivery");
    assert!(
        settled < e.created_at.unwrap(),
        "watermark {settled} is not behind the event it just delivered ({:?})",
        e.created_at
    );
    h.stop().await;
}
